### ISSUE TITLE
Float values in numeric columns are rounded up

### ISSUE THREAD

**Issue by dsuch** [tlocke/pg8000#15]
Hello,

I have found a case where floats sent to numeric columns get rounded up, which results in a loss of precision.

I can reproduce it under both pg8000 1.12.5 and 1.13.1. I am using PostgreSQL 10.3-1 under Ubuntu 18.04 64-bit.

Please consider a table as below:

```sql
create table abc(
  foo numeric(20, 7)
);
```

Now, let's add a float to it directly:

```python
from pg8000 import connect

# Connection details go here
kwargs = {}

conn = connect(**kwargs)
cursor = conn.cursor()

cursor.execute('insert into abc(foo) values(1560704964.331047)')
conn.commit()
cursor.execute('select foo from abc')
print('Result', cursor.fetchall())
conn.close()
```

This outputs:

```
Result ([Decimal('1560704964.3310470')],)
```

But let's add the same float as a template parameter now:

```python
from pg8000 import connect

# Connection details go here
kwargs = {}

conn = connect(**kwargs)
cursor = conn.cursor()

cursor.execute('insert into abc(foo) values(%s)', [1560704964.331047])
conn.commit()
cursor.execute('select foo from abc')
print('Result', cursor.fetchall())
conn.close()
```

The result is now:

```
Result ([Decimal('1560704964.3310500')],)
```

In other words, 1560704964.3310470 was rounded up to 1560704964.3310500.

My interpretation of the fact that in the first case the value was not rounded up is that PostgreSQL itself is not doing it. Otherwise, it likely would have done it no matter in what way the value is being provided to it.

I am now looking at pg8000's source code, in particular, its Struct.pack/unpack serialization methods to see if anything there can be related to it.

Unfortunately, I cannot use Decimal objects instead of floats though another idea is perhaps to convert the floats to string objects right before passing them to pg8000, which works around the rounding behaviour, but I would just like to understand what is actually happening and any pointers are appreciated.

Thanks.

**Comment by tlocke** [comment]
Hi @dsuch, leaving pg8000 out of the equation for a moment and just using `psql`:

```
postgres=# select 1560704964.331047;
     ?column?      
-------------------
 1560704964.331047
(1 row)
```
which is the expected answer, but if we do a cast:

```
postgres=# select cast(1560704964.331047 as double precision);
      float8      
------------------
 1560704964.33105
(1 row)
```

we get the rounding error that you noticed. Looking at pg8000:

```
import pg8000

conn = pg8000.connect(user="postgres", password="C.P.Snow", port=5438)
cursor = conn.cursor()

cursor.execute("create temporary table a ( foo numeric(20, 7));")
cursor.execute('insert into a(foo) values(%s)', [1560704964.331047])
conn.commit()

print(list(cursor.execute('select foo from a')))
# [[Decimal('1560704964.3310500')]]


cursor.execute("create temporary table b ( foo double precision);")
cursor.execute('insert into b(foo) values(%s)', [1560704964.331047])
conn.commit()

print(list(cursor.execute('select foo from b')))
# [[1560704964.331047]]

cursor.execute("create temporary table c ( foo numeric(20, 7));")
cursor.execute('insert into c(foo) values(%s)', ['1560704964.331047'])
conn.commit()

print(list(cursor.execute('select foo from c')))
# [[Decimal('1560704964.3310470')]]

conn.close()
```

So looking at these results I tentatively conclude that pg8000 is doing the right thing, and there's some rounding going on in PostgreSQL (which may well be the correct behavior). I realize that you can't use decimals, but that would be the ideal solution :-)

**Comment by tlocke** [comment]
I'm closing this for now, but please do re-open if you think pg8000 is doing the wrong thing.

**Comment by dsuch** [comment]
Hi @tlocke - yes, it made sense to close it. Let me just leave a note here that the workaround of using string objects does work nicely.
